### PR TITLE
[21295] Participant discovery structures refactor

### PR DESCRIPTION
**Pull request by elianalf**
## Description

This PR is a refactor of Participant discovery structures. In particular:

* Change the name of the enum `DISCOVERY_STATUS` → `ParticipantDiscoveryStatus`.
* Change the name of the callback: `onParticipantDiscovery` -> `on_participant_discovery`.
* Change the signature for both `RTPSParticipantListener` and `DomainParticipantListener`. The methods `on_participant_discover()` now take `ParticipantBuiltinTopicData` and a `ParticipantDiscoveryStatus`, instead of `ParticipantDiscoveryInfo`.
* Remove `ParticipantDiscoveryInfo`
* Extend `ParticipantBuiltinTopicData` with additional fields: locators, participant name, properties, lease duration
* Create an internal method to create `ParticipantBuiltinTopicData` from a `ParticipantProxyData`.
* Make `ParticipantProxyData` header private.
* Change signature of `fill_discovery_data_from_cdr_message()` to take ParticipantBuiltinTopicData

Related PR:
* Fast DDS Python PR: https://github.com/eProsima/Fast-DDS-python/pull/166
* Discovery Server https://github.com/eProsima/Discovery-Server/pull/99

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- N/A Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- ❌ Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- ❌ Changes are API compatible. 
- [X] New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 
    - Related documentation PR: eProsima/Fast-DDS-docs#874 -->
- N/A Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- _N/A_ If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: test/mock/rtps/ParticipantProxyData/rtps/builtin/data/ParticipantProxyData.hpp
Comment (MiguelCompany) [inline test/mock/rtps/ParticipantProxyData/rtps/builtin/data/ParticipantProxyData.hpp:1]: Please change the header guard in the mock

File: test/mock/rtps/RTPSParticipant/fastdds/rtps/participant/RTPSParticipant.hpp [hunk 1]
```diff
@@ -29,9 +29,10 @@
 #include <fastdds/dds/subscriber/qos/ReaderQos.hpp>
 #include <fastdds/fastdds_dll.hpp>
 #include <fastdds/rtps/attributes/RTPSParticipantAttributes.hpp>
-#include <fastdds/rtps/builtin/data/ParticipantProxyData.hpp>
+#include <fastdds/rtps/builtin/data/ParticipantBuiltinTopicData.hpp>
 #include <fastdds/rtps/common/Guid.hpp>
 
+#include <rtps/builtin/data/ParticipantProxyData.hpp>
```
Comment (MiguelCompany) [inline test/mock/rtps/RTPSParticipant/fastdds/rtps/participant/RTPSParticipant.hpp:35]:
```suggestion
```

File: test/mock/rtps/SecurityPluginFactory/rtps/security/MockAccessControlPlugin.h [hunk 2]
```diff
@@ -20,9 +20,9 @@
 #define FASTDDS_RTPS_SECURITY__MOCKACCESSCONTROLPLUGIN_H
 
 #include <fastdds/rtps/builtin/data/ReaderProxyData.hpp>
-#include <fastdds/rtps/builtin/data/ParticipantProxyData.hpp>
 
 #include <rtps/builtin/data/WriterProxyData.hpp>
+#include <rtps/builtin/data/ParticipantProxyData.hpp>
```
Comment (MiguelCompany) [inline test/mock/rtps/SecurityPluginFactory/rtps/security/MockAccessControlPlugin.h:25]:
```suggestion
#include <rtps/builtin/data/ParticipantProxyData.hpp>
#include <rtps/builtin/data/WriterProxyData.hpp>
```

File: test/mock/rtps/external_locators/rtps/network/utils/external_locators.hpp [hunk 3]
```diff
@@ -20,12 +20,12 @@
 #define FASTDDS_RTPS_NETWORK_UTILS_EXTERNAL_LOCATORS_HPP
 
 #include <fastdds/rtps/attributes/ExternalLocators.hpp>
-#include <fastdds/rtps/builtin/data/ParticipantProxyData.hpp>
 #include <fastdds/rtps/builtin/data/ReaderProxyData.hpp>
 #include <fastdds/rtps/common/LocatorList.hpp>
 #include <fastdds/rtps/common/LocatorSelectorEntry.hpp>
 
 #include <rtps/builtin/data/WriterProxyData.hpp>
+#include <rtps/builtin/data/ParticipantProxyData.hpp>
```
Comment (MiguelCompany) [inline test/mock/rtps/external_locators/rtps/network/utils/external_locators.hpp:28]:
```suggestion
#include <rtps/builtin/data/ParticipantProxyData.hpp>
#include <rtps/builtin/data/WriterProxyData.hpp>
```

File: versions.md [hunk 4]
```diff
@@ -90,6 +90,8 @@ Forthcoming
 * Setting vendor_id in the received CacheChange_t for Data and DataFrag.
 * Added new DynamicData to JSON serializer (`json_serialize`).
 * Added new DynamicType to IDL serializer (`idl_serialize`).
+* Refactor discovery info structures
+  * Participant discovery data refactor
```
Comment (MiguelCompany) [inline versions.md:94]: Remove this and add:
1. One line similar to line 20 in line 21
2. Two line in line 25 with `ParticipantProxyData` and `ParticipantDiscoveryInfo`

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: include/fastdds/rtps/builtin/data/ParticipantBuiltinTopicData.hpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ * @file ParticipantBuiltinTopicData.hpp
+ *
+ */
+
+#ifndef FASTDDS_RTPS_BUILTIN_DATA__PARTICIPANTBUILTINTOPICDATA_HPP
+#define FASTDDS_RTPS_BUILTIN_DATA__PARTICIPANTBUILTINTOPICDATA_HPP
+
+#include <fastdds/dds/builtin/topic/BuiltinTopicKey.hpp>
+#include <fastdds/dds/core/policy/QosPolicies.hpp>
+#include <fastdds/rtps/common/RemoteLocators.hpp>
```
[the remark this answers is not in the document]
Comment (elianalf) [reply include/fastdds/rtps/builtin/data/ParticipantBuiltinTopicData.hpp:25]: I am going to include them in alphabetic order:
#include <fastdds/dds/core/policy/QosPolicies.hpp>
#include <fastdds/rtps/builtin/data/BuiltinTopicKey.hpp>
#include <fastdds/rtps/common/RemoteLocators.hpp>

File: include/fastdds/statistics/dds/domain/DomainParticipant.hpp [hunk 6]
```diff
@@ -172,7 +171,7 @@ class DomainParticipant : public eprosima::fastdds::dds::DomainParticipant
      * @return RETCODE_ERROR if the  operation fails.
      */
     FASTDDS_EXPORTED_API eprosima::fastdds::dds::ReturnCode_t fill_discovery_data_from_cdr_message(
-            fastdds::rtps::ParticipantProxyData& data,
+            fastdds::rtps::ParticipantBuiltinTopicData& data,
```
[the remark this answers is not in the document]
Comment (elianalf) [reply include/fastdds/statistics/dds/domain/DomainParticipant.hpp:174]: I don't think I can remove this namespace

[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/statistics/dds/domain/DomainParticipant.hpp:174]: Then change it to `fastdds::dds`

**Review by MiguelCompany — APPROVED** [review]
LGTM with green CI

**Review by MiguelCompany — APPROVED** [review]
LGTM with green CI

**Review by MiguelCompany — APPROVED** [review]
LGTM with green CI

**Pull request merged**